##### Problem Statement 1
In each of the following situations, state whether it is a correctly stated hypothesis testing problem and why?
1. H0: μ = 25, H1: μ ≠ 25                                     
2. H0: σ > 10, H1: σ = 10                   
3. H0: x = 50, H1: x ≠ 50                                    
4. H0: p = 0.1, H1: p = 0.5                     
5. H0: s = 30, H1: s > 30

***Descriptions:***                  
1.Yes,                
Null Hypothesis H0 is 𝜇 = 25,                            
Alternative Hypothesis H1 is 𝜇 ≠ 25, which is ***Correct*** .                      
This is a two sided hypothesis. Null and Alternative hypothesis are contradicting with each other           
This is called a two-sided alternative hypothesis.                            

2.Incorrect,                 
Null Hypothesis H0 is 𝜎 > 10,                         
Alternative Hypothesis H1 is 𝜎 = 10, which is ***Incorrect*** .                    
As alternative hypothesis has to satisfy both the cases of 𝜎 = 10 and 𝜎 < 10.                 
If any of the cases matches then Null Hypothesis will be incorrect.                     
So alternative hypothesis H1 would be 𝜎 < = 10                      

3.Incorrect,                                           
Null Hypothesis H0 is 𝑥 = 50,                                     
Alternative Hypothesis H1 is 𝑥 ≠ 50, which is ***Incorrect*** .                                    
Hypothesis always talks about the population or distribution. Here it is a sample.                             

4.Incorrect                                            
Null Hypothesis H0 is 𝑝 = 0.1,                                        
Alternative Hypothesis H1 is 𝑝 = 0.5, which is ***Incorrect*** .                                      
As alternative hypothesis has to satisfy both the cases of 𝑝 < 0.1 and 𝑝 > 0.1.                     
So alternative hypothesis H1 would be 𝑝 ≠ 0.1                  
Either null or alternative should be true at any point of time. Here, if p=0.3 comes, then both null and alternative will fail.                      
                             
5.Incorrect                  
Null Hypothesis 𝐻0: 𝑠 = 30,                         
Alternative Hypothesis 𝐻1: 𝑠 > 30, which is **Incorrect**.                   
Hypothesis always talks about the population or distribution. Here it is a sample.              


###### Problem Statement 2:
The college bookstore tells prospective students that the average cost of its
textbooks is Rs. 52 with a standard deviation of Rs. 4.50. A group of smart statistics
students thinks that the average cost is higher. To test the bookstore’s claim against
their alternative, the students will select a random sample of size 100. Assume that
the mean from their random sample is Rs. 52.80. Perform a hypothesis test at the
5% level of significance and state your decision.

In [1]:
import numpy as math
from scipy import stats

pMean = 52 # populatin mean
pStd = 4.50 # population standard deviation
Sample_size = 100 # sample size
sMean = 52.80 # sample mean
signi_level = 0.05 # significance level

# Hypothesis H0: pMean = 52, H1: pMean ≠ 52
# This will be a two-tailed test
    
def Z_test(sMean, pMean,Sample_size,pStd ):  # Z test calculation
    SE =(pStd)/math.sqrt(Sample_size) # Standard error
    Z_val = (sMean - pMean) / SE
    return Z_val

val = Z_test(sMean, pMean, Sample_size,pStd)
print('The computed value of Z is: ',val)

# Now we need to find out the critical value for significance level 5% = 0.05
p=1-signi_level/2
z_critical=stats.norm.ppf(p)
print('The value of z_critical is: ',z_critical)

def check_test(val,z_critical):
    if val > z_critical:
        print("Null Hypothesis Rejected")
    else:
        print("Null Hypothesis Accepted")
check_test(val,z_critical)

The computed value of Z is:  1.7777777777777715
The value of z_critical is:  1.959963984540054
Null Hypothesis Accepted


****Conclusions****
The criticial value of Z is: Z = +1.96 and -1.96 at 5% level of significance.                   
Our computed value of Z is: Z = 1.78.                  

So, computed value falls in acceptance region. Hence we accept the Null Hypothesis.          
Hence the mean average cost of text book is: $52.    

###### Problem Statement 3:
A certain chemical pollutant in the Genesee River has been constant for several
years with mean μ = 34 ppm (parts per million) and standard deviation σ = 8 ppm. A
group of factory representatives whose companies discharge liquids into the river is
now claiming that they have lowered the average with improved filtration devices. A
group of environmentalists will test to see if this is true at the 1% level of
significance. Assume \ that their sample of size 50 gives a mean of 32.5 ppm.
Perform a hypothesis test at the 1% level of significance and state your decision.

In [2]:
import numpy as math
from scipy import stats

pMean = 34 # populatin mean
pStd = 8 # population standard deviation
Sample_size = 50 # sample size
sMean =32.5 # sample mean
signi_level = 0.01 # significance level

# Hypothesis: 
# H0: pMean = 34, 
# H1: pMean ≠ 34
# This will be a two-tailed test

    
def Z_test(sMean, pMean,Sample_size,pStd ):  # Z test calculation
    SE =(pStd)/math.sqrt(Sample_size) # Standard error
    Z_val = (sMean - pMean) / SE
    return Z_val

val = Z_test(sMean, pMean, Sample_size,pStd)
print('The computed value of Z is: ',val)

# Now we need to find out the critical value for significance level 1% = 0.01

p=1-signi_level/2
z_critical=stats.norm.ppf(p)
print('The value of z_critical is: ',z_critical)

def check_test(val,z_critical):
    if val > z_critical:
        print("Null Hypothesis Rejected")
    else:
        print("Null Hypothesis Accepted")
check_test(val,z_critical)


The computed value of Z is:  -1.3258252147247767
The value of z_critical is:  2.5758293035489004
Null Hypothesis Accepted


****Conclusions****       
The criticial value of Z is: Z = +2.58 and -2.58 at 1% level of significance.                             
Our computed value of Z is: Z = -1.33.                            

So, computed value falls in acceptance region. Hence we accept the Null Hypothesis.                        
Hence there claim that they have lowered the average discharge with improved filteration devices is true.                 

##### Problem Statement 4:
Based on population figures and other general information on the U.S. population,
suppose it has been estimated that, on average, a family of four in the U.S. spends
about $1135 annually on dental expenditures. Suppose further that a regional dental
association wants to test to determine if this figure is accurate for their area of
country. To test this, 22 families of 4 are randomly selected from the population in
that area of the country and a log is kept of the family’s dental expenditure for one
year. The resulting data are given below. Assuming, that dental expenditure is
normally distributed in the population, use the data and an alpha of 0.5 to test the
dental association’s hypothesis.
1008, 812, 1117, 1323, 1308, 1415, 831, 1021, 1287, 851, 930, 730, 699,
872, 913, 944, 954, 987, 1695, 995, 1003, 994

In [3]:
import numpy as mat
from scipy import stats
import statistics

pMean = 1135 # populatin mean
Sample_size = 22 # sample size
lst=[1008, 812, 1117, 1323, 1308, 1415, 831, 1021, 1287, 851, 930, 730, 699, 872, 913, 944, 954, 987, 1695, 995, 1003 ,994]
sMean =math.mean(lst) # sample mean
pStd =statistics.stdev(lst)# population standard deviation
signi_level = 0.05 # significance level

# Hypothesis: 
# H0: pMean = 1135, 
# H1: pMean ≠ 1135
# This will be a two-tailed test

def Z_test(sMean, pMean,Sample_size,pStd ):  # Z test calculation
    SE =(pStd)/math.sqrt(Sample_size) # Standard error
    Z_val = (sMean - pMean) / SE
    return Z_val

val = Z_test(sMean, pMean, Sample_size,pStd)
print('The computed value of Z is: ',val)

# Now we need to find out the critical value for significance level = 0.05

p=1-signi_level/2
print(p)
z_critical=stats.norm.ppf(p)
print('The value of z_critical is: ',z_critical)

def check_test(val,z_critical):
    if val > z_critical:
        print("Null Hypothesis Rejected")
    else:
        print("Null Hypothesis Accepted")
check_test(val,z_critical)

The computed value of Z is:  -2.023137479931484
0.975
The value of z_critical is:  1.959963984540054
Null Hypothesis Accepted


##### Conclusions
The criticial value of Z is: Z = +1.96 and -1.96 at 5% level of significance.                     
Our computed value of Z is: Z = -2.023.                              

So, computed value falls in rejection region. Hence we reject the Null Hypothesis.                   
    Hence the average dental expenses is not accurate for this area                         

###### Problem Statement 5:
In a report prepared by the Economic Research Department of a major bank the Department manager maintains that the average annual family income on Metropolis is dollar 48,432. What do you conclude about the validity of the report if a random sample of 400 families shows and average income of $48,574 with a standard deviation of 2000?

In [4]:
import numpy as math
from scipy import stats

pMean = 48432 # populatin mean
pStd = 2000 # population standard deviation
Sample_size = 400 # sample size
sMean =48574 # sample mean

# Hypothesis: 
# H0: pMean = 48432, 
# H1: pMean ≠ 48432
# This will be a two-tailed test

def Z_test(sMean, pMean,Sample_size,pStd ):  # Z test calculation
    SE =(pStd)/math.sqrt(Sample_size) # Standard error
    Z_val = (sMean - pMean) / SE
    return Z_val

val = Z_test(sMean, pMean, Sample_size,pStd)
print('The computed value of Z is: ',val)

The computed value of Z is:  1.42


###### Problem Statement 6:
Suppose that in past years the average price per square foot for warehouses in the
United States has been dollar 32.28. A national real estate investor wants to determine
whether that figure has changed now. The investor hires a researcher who randomly
samples 19 warehouses that are for sale across the United States and finds that the
mean price per square foot is dollar 31.67, with a standard deviation of $1.29. assume
that the prices of warehouse footage are normally distributed in population. If the
researcher uses a 5% level of significance, what statistical conclusion can be
reached? What are the hypotheses?

In [5]:
import numpy as math
from scipy import stats

pMean = 32.28 # populatin mean
pStd = 1.29 # population standard deviation
Sample_size = 19 # sample size
sMean =31.67 # sample mean
signi_level = 0.05 # significance level

# Hypothesis: 
# H0: pMean = 32.28, 
# H1: pMean ≠ 32.28
# This will be a two-tailed test

def Z_test(pMean,pStd,Sample_size,sMean):
    SE=pStd/math.sqrt(Sample_size)
    Z_val=pMean-sMean/SE
    return Z_val
    
val=Z_test(pMean,pStd,Sample_size,sMean)
print("The computed value of Z", val)

# Now we need to find out the critical value for significance level 5% = 0.05
p=1-signi_level/2
print(p)
Z_critical=stats.norm.ppf(p)
print('The value of z_critical is: ',Z_critical)

def check_test(val,Z_critical):
    if val>Z_critical:
        print("Null Hypothesis Rejected")
    else:
        print("Null Hypothesis Accepted")
check_test(val,Z_critical)

The computed value of Z -74.73265855963811
0.975
The value of z_critical is:  1.959963984540054
Null Hypothesis Accepted


##### Conclusions
The criticial value of Z is: Z = +1.96 and -1.96 at 5% level of significance.              
Our computed value of Z is: Z = -2.061.                

So, computed value falls in rejection region. Hence we reject the Null Hypothesis.                   
Hence the average price per square foot for warehouses has changed now.              

##### Problem Statement 7:
Fill in the blank spaces in the table and draw your conclusions from it

 
| Acceptance Region | Sample Size | α | β at μ = 52 | β at μ = 50.5| 
| ---- | ---- | ---- | ---- | ---- |
| 48.5 <x̅ < 51.5 | 10 |  |  |  |
| 48 <x̅ < 52 | 10 |  |  |  |
| 48.81 <x̅ < 51.9 | 16 |  |  |  |
| 48.42 <x̅ < 51.58 | 16 |  |  |  |
					

In [6]:
import math
from scipy.stats import norm
# α = P(type I error) = P(reject H0 when H0 is true)

# let say 𝜇 = 50, 𝜎 = 2.5
# α = P(48.5 <x̅ < 51.5) =  P(x < 48.5 when𝜇 = 50) + P(x > 51.5 when𝜇 = 50)

z1 = (48.5-50)/ (2.5/math.sqrt(10)) # applying central limit theroem
z2 = (51.5-50)/ (2.5/math.sqrt(10))
# α = P(Z < z1) + P(Z>z2)
α = norm.cdf(z1) +(1 - norm.cdf(z2))
print('Calculated α value is: ', round(α, 4))

# β = P(48.5 <x̅ < 51.5) at 𝜇 = 52 # type II error
# β = P(((48.5-52)/(2.5/math.sqrt(10)) < Z < (51.5-52)/((2.5/math.sqrt(10)))

z3 = (48.5-52)/ (2.5/math.sqrt(10))
z4 = (51.5-52)/(2.5/math.sqrt(10))
# β = P(Z < z4) - P(Z<z3)
β = norm.cdf(z4) - norm.cdf(z3)
print('Calculated β value at 𝜇 = 52 is: ', round(β, 4))


# β = P(48.5 <x̅ < 51.5) at 𝜇 = 50.5 # type II error
# β = P(((48.5-50.5)/(2.5/math.sqrt(10)) < Z < (51.5-50.5)/((2.5/math.sqrt(10)))

z5 = (48.5-50.5)/ (2.5/math.sqrt(10))
z6 = (51.5-50.5)/(2.5/math.sqrt(10))
# β = P(Z < z6) - P(Z<z5)
β = norm.cdf(z6) - norm.cdf(z5)
print('Calculated β value at 𝜇 = 50.5 is: ', round(β, 4))

Calculated α value is:  0.0578
Calculated β value at 𝜇 = 52 is:  0.2635
Calculated β value at 𝜇 = 50.5 is:  0.8913


In [7]:
# α = P(type I error) = P(reject H0 when H0 is true)

# let say 𝜇 = 50, 𝜎 = 2.5
# α = P(48 <x̅ < 52) =  P(x < 48 when𝜇 = 50) + P(x > 52 when𝜇 = 50)

z1 = (48-50)/ (2.5/math.sqrt(10)) # applying central limit theroem
z2 = (52-50)/ (2.5/math.sqrt(10))
# α = P(Z < z1) + P(Z>z2)
α = norm.cdf(z1) +(1 - norm.cdf(z2))
print('Calculated α value is: ', round(α, 4))

# β = P(48 <x̅ < 52) at 𝜇 = 52 # type II error
# β = P(((48-52)/(2.5/math.sqrt(10)) < Z < (52-52)/((2.5/math.sqrt(10)))

z3 = (48-52)/ (2.5/math.sqrt(10))
z4 = (52-52)/(2.5/math.sqrt(10))
# β = P(Z < z4) - P(Z<z3)
β = norm.cdf(z4) - norm.cdf(z3)
print('Calculated β value at 𝜇 = 52 is: ', round(β, 4))


# β = P(48 <x̅ < 52) at 𝜇 = 50.5 # type II error
# β = P(((48-50.5)/(2.5/math.sqrt(10)) < Z < (52-50.5)/((2.5/math.sqrt(10)))

z5 = (48-50.5)/ (2.5/math.sqrt(10))
z6 = (52-50.5)/(2.5/math.sqrt(10))
# β = P(Z < z6) - P(Z<z5)
β = norm.cdf(z6) - norm.cdf(z5)
print('Calculated β value at 𝜇 = 50.5 is: ', round(β, 4))

Calculated α value is:  0.0114
Calculated β value at 𝜇 = 52 is:  0.5
Calculated β value at 𝜇 = 50.5 is:  0.9703


In [8]:
# α = P(type I error) = P(reject H0 when H0 is true)

# let say 𝜇 = 50, 𝜎 = 2.5
# α = P(48.81 <x̅ < 51.19) =  P(x < 48.81 when𝜇 = 50) + P(x > 51.19 when𝜇 = 50)

z1 = (48.81-50)/ (2.5/math.sqrt(16)) # applying central limit theroem
z2 = (51.19-50)/ (2.5/math.sqrt(16))
# α = P(Z < z1) + P(Z>z2)
α = norm.cdf(z1) +(1 - norm.cdf(z2))
print('Calculated α value is: ', round(α, 4))

# β = P(48.81 <x̅ < 51.19) at 𝜇 = 52 # type II error
# β = P(((48.81-52)/(2.5/math.sqrt(16)) < Z < (51.19-52)/((2.5/math.sqrt(16)))

z3 = (48.81-52)/ (2.5/math.sqrt(16))
z4 = (51.19-52)/(2.5/math.sqrt(16))
# β = P(Z < z4) - P(Z<z3)
β = norm.cdf(z4) - norm.cdf(z3)
print('Calculated β value at 𝜇 = 52 is: ', round(β, 4))


# β = P(48.81 <x̅ < 51.19) at 𝜇 = 50.5 # type II error
# β = P(((48.81-50.5)/(2.5/math.sqrt(16)) < Z < (51.19-50.5)/((2.5/math.sqrt(16)))

z5 = (48.81-50.5)/ (2.5/math.sqrt(16))
z6 = (51.19-50.5)/(2.5/math.sqrt(16))
# β = P(Z < z6) - P(Z<z5)
β = norm.cdf(z6) - norm.cdf(z5)
print('Calculated β value at 𝜇 = 50.5 is: ', round(β, 4))

Calculated α value is:  0.0569
Calculated β value at 𝜇 = 52 is:  0.0975
Calculated β value at 𝜇 = 50.5 is:  0.8618


In [9]:
# α = P(type I error) = P(reject H0 when H0 is true)

# let say 𝜇 = 50, 𝜎 = 2.5
# α = P(48.42 <x̅ < 51.58) =  P(x < 48.42 when𝜇 = 50) + P(x > 51.58 when𝜇 = 50)

z1 = (48.42-50)/ (2.5/math.sqrt(16)) # applying central limit theroem
z2 = (51.58-50)/ (2.5/math.sqrt(16))
# α = P(Z < z1) + P(Z>z2)
α = norm.cdf(z1) +(1 - norm.cdf(z2))
print('Calculated α value is: ', round(α, 4))

# β = P(48.42 <x̅ < 51.58) at 𝜇 = 52 # type II error
# β = P(((48.42-52)/(2.5/math.sqrt(16)) < Z < (51.58-52)/((2.5/math.sqrt(16)))

z3 = (48.42-52)/ (2.5/math.sqrt(16))
z4 = (51.58-52)/(2.5/math.sqrt(16))
# β = P(Z < z4) - P(Z<z3)
β = norm.cdf(z4) - norm.cdf(z3)
print('Calculated β value at 𝜇 = 52 is: ', round(β, 4))


# β = P(48.81 <x̅ < 51.19) at 𝜇 = 50.5 # type II error
# β = P(((48.81-50.5)/(2.5/math.sqrt(16)) < Z < (51.19-50.5)/((2.5/math.sqrt(16)))

z5 = (48.42-50.5)/ (2.5/math.sqrt(16))
z6 = (51.58-50.5)/(2.5/math.sqrt(16))
# β = P(Z < z6) - P(Z<z5)
β = norm.cdf(z6) - norm.cdf(z5)
print('Calculated β value at 𝜇 = 50.5 is: ', round(β, 4))

Calculated α value is:  0.0115
Calculated β value at 𝜇 = 52 is:  0.2508
Calculated β value at 𝜇 = 50.5 is:  0.9576


##### updated table - 
 
| Acceptance Region | Sample Size | α | β at μ = 52 | β at μ = 50.5| 
| ---- | ---- | ---- | ---- | ---- |
| 48.5 <x̅ < 51.5 | 10 |0.0578|0.2635|0.8913|
| 48 <x̅ < 52 | 10 | 0.0114 |0.5  | 0.9703 |
| 48.81 <x̅ < 51.9 | 16 | 0.0569 |0.0975  |  0.8618|
| 48.42 <x̅ < 51.58 | 16 | 0.0115 | 0.2508 | 0.9576 |

##### Problem Statement 8:
Find the t-score for a sample size of 16 taken from a population with mean 10 when
the sample mean is 12 and the sample standard deviation is 1.5.

In [10]:
pMean = 10 # population mean
sMean = 12 # sample mean
sSize = 16 # sample size
sStd = 1.5 # sample standard deviation

# Function for calculating t-score

def t_score(pMean, sMean, sStd, sSize):
    se = sStd / math.sqrt(sSize)
    score = (sMean - pMean) / se
    return score
val=t_score(pMean, sMean, sStd, sSize)
print("The value of t-score is :", val)

The value of t-score is : 5.333333333333333


###### Problem Statement 9:
Find the t-score below which we can expect 99% of sample means will fall if samples
of size 16 are taken from a normally distributed population.

In [11]:
from scipy.stats import t
#𝑑𝑓 = N − 1

alpha = 0.01 # significance level (1- 0.99) 
df = 15 # degree of freedom (16-1)
cv = t.ppf(1.0 - alpha, df) # critical value
print('Critical t-score value is: ', cv)

Critical t-score value is:  2.602480294995493


##### Problem Statement 10:
If a random sample of size 25 drawn from a normal population gives a mean of 60 and a standard deviation of 4, find the range of t-scores where we can expect to find the middle 95% of all sample means. Compute the probability that (−𝑡0.05 <𝑡<𝑡0.10).

In [12]:
from scipy.stats import t

sMean = 60 # sample mean
sSize = 25 # sample size
sStd = 4 # sample standard deviation

df = 24 # degree of freedoms = 25-1(N-1)

# s = 0.05, 5% significance level
# for a two-tailed test alpha =s / 2 = 0.05/2 = 0.025

alpha = 0.025

cv = t.ppf(1.0 - alpha, df) # critical value
print('Critical t-score value is: ', cv)


Critical t-score value is:  2.0638985616280205


###### Compute the probability that (−𝑡0.05 <𝑡<𝑡0.10)
P((−𝑡0.05 <𝑡 <𝑡0.10)) represents an area of 0.10 at the right tail and an area of 0.05 at the left tail.            
P((−𝑡0.05 <𝑡 <𝑡0.10)) = (1 - 0.10 - 0.05) = 0.85                 

So, computed probaility is: 0.85              

##### Problem Statement 11:
Two-tailed test for difference between two population means
Is there evidence to conclude that the number of people travelling from Bangalore to
Chennai is different from the number of people travelling from Bangalore to Hosur in
a week, given the following:                           
Population 1: Bangalore to Chennai n1 = 1200                                                  
x1 = 452                                                             
s1 = 212                                                   
Population 2: Bangalore to Hosur n2 = 800                          
x2 = 523                                       
s2 = 185                                                        

In [13]:
import math
from scipy import stats

# H0: 𝜇1 - 𝜇2 = 0 # null hypothesis

# H0: 𝜇1 = 𝜇2
# H1: 𝜇1 ≠ 𝜇2

x1 = 452
x2 = 523
s1 = 212
n1 = 1200
s2 = 185
n2 = 800

def standard_error(s1, n1, s2, n2):
    stdErr = math.sqrt(((s1)**2/n1)+((s2)**2/n2))
    return stdErr

SE = standard_error(s1, n1, s2, n2)
print('Computed Standard Error is: ', SE)

def Z_test(x1, x2, SE):  # Z test calculation
    Z_val = (x1 - x2) / SE
    return Z_val

Z_score = Z_test(x1, x2, SE)
print('Computed value of Z is: ', Z_score)

# Considering significance level 5% = 0.05
alpha=0.05
p=1-alpha/2
print(p)
Z_critical=stats.norm.ppf(p)
print('The value of z_critical is: ',Z_critical)

def check_test(Z_score,Z_critical):
    if Z_score>Z_critical:
        print("Null Hypothesis Rejected")
    else:
        print("Null Hypothesis Accepted")
check_test(Z_score,Z_critical)

Computed Standard Error is:  8.957375917830698
Computed value of Z is:  -7.926428526759299
0.975
The value of z_critical is:  1.959963984540054
Null Hypothesis Accepted


###### Conclusions
The criticial value of Z is: Z = +1.96 and -1.96 at 5% level of significance.
Our computed value of Z is: Z = -7.926.

So, computed value falls in rejection region. Hence we reject the Null Hypothesis.
Hence the mean of the two population are different.

###### Problem Statement 12:
Is there evidence to conclude that the number of people preferring Duracell battery is different from the number of people preferring Energizer battery, given the following:
Population 1: Duracell                   
n1 = 100                     
x1 = 308                              
s1 = 84                                  
Population 2: Energizer                                      
n2 = 100                                  
x2 = 254                                    
s2 = 67                                          

In [14]:
import math
from scipy import stats
# H0: 𝜇1 - 𝜇2 = 0 # null hypothesis

# H0: 𝜇1 = 𝜇2
# H1: 𝜇1 ≠ 𝜇2

n1 = 100                     
x1 = 308                              
s1 = 84
n2 = 100                                  
x2 = 254                                    
s2 = 67

def standard_error(s1, n1, s2, n2):
    stdErr = math.sqrt(((s1)**2/n1)+((s2)**2/n2))
    return stdErr

SE = standard_error(s1, n1, s2, n2)
print('Computed Standard Error is: ', SE)

def Z_test(x1, x2, SE):  # Z test calculation
    Z_val = (x1 - x2) / SE
    return Z_val

Z_score = Z_test(x1, x2, SE)
print('Computed value of Z is: ', Z_score)

# Considering significance level 5% = 0.05
alpha=0.05
p=1-alpha/2
print(p)
Z_critical=stats.norm.ppf(p)
print('The value of z_critical is: ',Z_critical)

def check_test(Z_score,Z_critical):
    if Z_score>Z_critical:
        print("Null Hypothesis Rejected")
    else:
        print("Null Hypothesis Accepted")
check_test(Z_score,Z_critical)

Computed Standard Error is:  10.74476616776745
Computed value of Z is:  5.025702668336442
0.975
The value of z_critical is:  1.959963984540054
Null Hypothesis Rejected


##### Conclusions
The criticial value of Z is: Z = +1.96 and -1.96 at 5% level of significance.                  
Our computed value of Z is: Z = 5.026.                        

So, computed value falls in rejection region. Hence we reject the Null Hypothesis.                
Hence the number of people preferring Duracel battery is different than the number of people preferring Energizer.

###### Problem Statement 13:
Pooled estimate of the population variance Does the data provide sufficient evidence to conclude that average percentage increase in the price of sugar differs when it is sold at two different prices?
Population 1:                                 
Price of sugar = Rs. 27.50                   
n1 = 14                          
x1 = 0.317%                           
s1 = 0.12%                            
Population 2:                             
Price of sugar = Rs. 20.00                   
n2 = 9                       
x2 = 0.21%                               
s2 = 0.11%                         

In [15]:
import math
from scipy import stats
from scipy.stats import t

x1 = 0.317
x2 = 0.21
s1 = 0.12
n1 = 14
s2 = 0.11
n2 = 9

# H0: 𝜇1 = 𝜇2
# H1: 𝜇1 ≠ 𝜇2

Sp = math.sqrt((((n1-1)*(s1)**2) +  ((n2-1)*(s2)**2)) / (n1+n2-2))  # the pooled estimate of the common standard deviaton

def standard_error(Sp, n1, n2):
    stdErr = Sp * math.sqrt((1/n1)+(1/n2))
    return stdErr

SE = standard_error(Sp,n1,n2)
print('Computed Standard Error is: ', SE)

def Z_test(x1, x2, SE):  # Z test calculation
    Z_val = (x1 - x2) / SE
    return Z_val

Z_score = Z_test(x1, x2, SE)
print('Computed value of Z is: ', Z_score)

# Considering significance level 5% = 0.05

# calculate the critical value
#df = n1+n2-2 = 14+9-2 = 21 -> degree of freedom
# this will be a two-tailed test
# alpha=0.05/2 =0.025

alpha = 0.025 
df = 21
cv = t.ppf(1.0 - alpha, df) # critical value
print('Critical t-score value is: ', cv)

Computed Standard Error is:  0.04968532879234398
Computed value of Z is:  2.15355322387416
Critical t-score value is:  2.079613844727662


###### Conclusions
The criticial value of t is: t = 2.08 at 5% level of significance.            
Our computed value of t is: t = 2.15.                   

Decision rule is reject H0 if t_critical <= t_computed. Here 2.08 < 2.15, Hence we may reject the Null Hypothesis at 5% level of significance.                      
So, average percentage increase of price of sugar will differ.            

##### Problem Statement 14:
The manufacturers of compact disk players want to test whether a small price reduction is enough to increase sales of their product. Is there evidence that the small price reduction is enough to increase sales of compact disk players?                                
Population 1: Before reduction                                    
n1 = 15                                             
x1 = Rs. 6598                                                
s1 = Rs. 844                                                    
Population 2: After reduction                                      
n2 = 12                                                                            
x2 = RS. 6870                                                                        
s2 = Rs. 669                                                                                 

In [16]:
import math
from scipy import stats
from scipy.stats import t

# H0: 𝜇1 = 𝜇2
# H1: 𝜇1 ≠ 𝜇2

x1 = 6598
x2 = 6870
s1 = 844
n1 = 15
s2 = 669
n2 = 12

Sp = math.sqrt((((n1-1)*(s1)**2) +  ((n2-1)*(s2)**2)) / (n1+n2-2))  # the pooled estimate of the common standard deviaton

def standard_error(Sp, n1, n2):
    stdErr = Sp * math.sqrt((1/n1)+(1/n2))
    return stdErr

SE = standard_error(Sp,n1,n2)
print('Computed Standard Error is: ', SE)

def Z_test(x1, x2, SE):  # Z test calculation
    Z_val = (x2 - x1) / SE
    return Z_val

Z_score = Z_test(x1, x2, SE)
print('Computed value of Z is: ', Z_score)

# calculate the critical value
#df = n1+n2-2 = 15+12-2 = 25 -> degree of freedom
# this will be a two-tailed test

alpha = 0.1 #let say significance level 10% 
df = 25
cv = t.ppf(1.0 - alpha, df) # critical value
print('Critical t-score value is: ', cv)

Computed Standard Error is:  298.9569366982476
Computed value of Z is:  0.9098300343990459
Critical t-score value is:  1.3163450765592588


###### Conclusions
The criticial value of t is: t = +1.316 at 10% level of significance.            
Our computed value of t is: t = 0.91.         

Decision rule is reject H0 if t_critical <= t_computed. Here 1.316 not <= 0.91, Hence we may not reject the Null Hypothesis.
So the small price reduction is not enough to increase sales of compact disk players.

###### Problem Statement 15:
Comparisons of two population proportions when the hypothesized difference is zero Carry out a two-tailed test of the equality of banks’ share of the car loan market in 1980 and 1995.                   
Population 1: 1980                        
n1 = 100                     
x1 = 53                    
𝑝 1 = 0.53                 
Population 2: 1985                    
n2 = 100                    
x2 = 43                                  
𝑝 2= 0.43                               

In [17]:
import math
from scipy.stats import norm

n1 = 100 # sample size
x1 = 53 # number of success in sample n1
p1 = 0.53 # sample1 proportion

n2 = 100 # sample size
x2 = 43 # number of success in sample n2
p2 = 0.43 # sample2 proportion


# H0: p1 = p2 -> null hypothesis
# H1: p1 ≠ p2 -> alternate hypothesis

p_hat = (x1+x2) / (n1+n2) # pooled proportion
print("p_hat is:" ,p_hat)

def standard_error(p_hat, n1, n2):
    std_error = math.sqrt(((p_hat/n1)*(1-p_hat)) + ((p_hat/n2)*(1-p_hat)))
    return std_error

se = standard_error(p_hat, n1, n2)
print('Calculated standard error is: ',se)

# calculate Z-test statistics

def Z_score(p1, p2, se):
    Z_test = (p1-p2) / se
    return Z_test
Z_val = Z_score(p1, p2, se)
print('Calculated Z test value is: ',Z_val)

# Considering significance level 10% = 0.1
# So, Alpha = 0.1
# As two-tailed test then, Alpha/2 = 0.05
# probability = 1 - alpha/2 = 1-0.05 = 0.95
p = 0.95
value = norm.ppf(p) # returns observaton values for provided probability which is less than or equal to the provided probability
print('The critical value for Z is: ', value)
# confirm with cdf
p = norm.cdf(value) # Returns the probability for an observation equal to or lesser than a specific value from the distribution
print(p)

p_hat is: 0.48
Calculated standard error is:  0.07065408693062278
Calculated Z test value is:  1.4153462926807452
The critical value for Z is:  1.6448536269514722
0.95


**Conclusions**                  
The criticial value of Z is: Z = -1.645 and + 1.645 at 10% level of significance.            
Our computed value of Z is: Z = 1.415                                

So, computed value falls in acceptance region. Hence we accept the Null Hypothesis.
Hence two population proportions are equal.

##### Problem Statement 16:                                 
Carry out a one-tailed test to determine whether the population proportion of traveler’s check buyers who buy at least $2500 in checks when sweepstakes prizes are offered as at least 10% higher than the proportion of such buyers when no sweepstakes are on.
Population 1: With sweepstakes                        
n1 = 300                   
x1 = 120                
𝑝1 = 0.40                   
Population 2: No sweepstakes                  
n2 = 700                   
x2 = 140                 
𝑝2 = 0.20                    

In [18]:
import math
from scipy.stats import norm

n1 = 300 # sample size
x1 = 120 # number of success in sample n1
p1 = 0.40 # sample1 proportion

n2 = 700 # sample size
x2 = 140 # number of success in sample n2
p2 = 0.20 # sample2 proportion
D = 0.10

# H0: p1 - p2 <= 0.1 -> null hypothesis
# H1: p1 - p2 > 0.1 -> alternate hypothesis

p_hat = (x1+x2) / (n1+n2) # pooled proportion
print("p_hat is:" ,p_hat)

def standard_error(p_hat, n1, n2):
    std_error = math.sqrt(((p_hat/n1)*(1-p_hat)) + ((p_hat/n2)*(1-p_hat)))
    return std_error

se = standard_error(p_hat, n1, n2)
print('Calculated standard error is: ',se)

# calculate Z-test statistics

def Z_score(p1, p2, D, se):
    Z_test = (p1-p2-D) / se
    return Z_test
Z_val = Z_score(p1, p2, D, se)
print('Calculated Z test value is: ',Z_val)

# Considering significance level 10% = 0.1
# So, Alpha = 0.1
# As one-tailed test then, Alpha = 0.1
# probability = 1 - 0.1 = 0.90
p = 0.90
value = norm.ppf(p) # returns observaton values for provided probability which is less than or equal to the provided probability
print('The critical value for Z is: ', value)
# confirm with cdf
p = norm.cdf(value) # Returns the probability for an observation equal to or lesser than a specific value from the distribution
print(p)

p_hat is: 0.26
Calculated standard error is:  0.030268638492513603
Calculated Z test value is:  3.3037495236111525
The critical value for Z is:  1.2815515655446004
0.8999999999999999


##### Conclusions
The criticial value of Z is: Z = 1.2815 at 10% level of significance.
Our computed value of Z is: Z = 3.30375

So, computed value falls in rejection region. Hence we reject the Null Hypothesis.
Hence alternate hypothesis is true here.

##### Problem Statement 17:
A die is thrown 132 times with the following results: Number turned up: 1, 2, 3, 4, 5, 6 Frequency: 16, 20, 25, 14, 29, 28 Is the die unbiased? Consider the degrees of freedom as 𝑝 − 1.

In [19]:
from scipy.stats import chi2

# H0: The die is unbiased
# H1: The die is biased

# In case of null hypothesis, the die would be unbiased if the frequency of each number is equal.
# frequency Expected = 132/6 = 22 for each number.

O = [16, 20, 25, 14, 29, 28] # Observed_frequency
E = 22 # Expected_frequency for each number

def chiSquare_Val(O, E):
    sum = 0
    for num in O:
        sum += (num-E)**2
    
    chi_val = sum / E
    return chi_val

X_sqr = chiSquare_Val(O, E)

print('Calculated Chi-Square value is: ', X_sqr)


# Considering significace level ∝ = 5% = 0.05
# degree of freedom = p-1 = 6-1=5

# prob = 1-0.05 = 0.95
p = 0.95
dof = 5

critical = chi2.ppf(p, dof)
print('Critical value of Chi-Square is: ', critical)

Calculated Chi-Square value is:  9.0
Critical value of Chi-Square is:  11.070497693516351


##### Conclusions
The criticial value of Chi-Square is: Χ^2 = 11.07 at 5% level of significance.
Our computed value of Chi-Square is: Χ^2 = 9.0

So, computed value falls in acceptance region. Hence we accept the Null Hypothesis.
Hence the die is unbiased.

#### Problem Statement 18: 
In a certain town, there are about one million eligible voters. A simple random sample of 10,000 eligible voters was chosen to study the relationship between gender and participation in the last election. The results are summarized in the following 2X2 (read two by two) contingency table:

 
| | Men | Women | 
| ---- | ---- | ---- |
| Voted | 2792| 3591 |
| Not Voted| 1486 | 2131 |


We would want to check whether being a man or a woman (columns) is independent of having voted in the last election (rows). In other words, is “gender and voting independent”? 

In [20]:
import pandas as pd
from scipy.stats import chi2

# creating view for Observed table

voter = {'Men': [2792, 1486], 'Women': [3591, 2131]}
df = pd.DataFrame(voter, columns = ['Men', 'Women'], index = ['Voted', 'Not Voted'])
print('Observed table: ')
print(df,"\n")

# H0: Gender and voting is independent --> Null hypothesis
#H1 : Gender and voting is not independent --> Alternate hypothesis

# Now we need to create the 'Expected table' for the mentioned null hypothesis

# Total men and women / Total voted and not voted
Total_Voters_sample = 10000

# Total men voted and not voted
T_men = df.sum(axis=0)[0]

# Total women voted and not voted
T_women = df.sum(axis=0)[1]

# Total voted (Men + Women)
T_voted = df.sum(axis=1)[0]

# Total not voted (Men + Women)
T_notVoted = df.sum(axis=1)[1]

# Creating data for Expected table
Men_Voted = round((T_voted / Total_Voters_sample)* T_men) # rounding off the numbers
Men_NotVoted = round((T_notVoted / Total_Voters_sample) * T_men )
Women_Voted = round((T_voted / Total_Voters_sample) * T_women)
Women_NotVoted = round((T_notVoted / Total_Voters_sample) * T_women)


# Creating view for Expected table

voter = {'Men': [Men_Voted, Men_NotVoted], 'Women': [Women_Voted, Women_NotVoted]}
df_exp = pd.DataFrame(voter, columns = ['Men', 'Women'], index = ['Voted', 'Not Voted'])
print('Expected table: ')
print(df_exp,"\n")


# Now applying Chi-Square independence formula on Observed and Expected table

s1 = ((df['Men'][0] - df_exp['Men'][0])**2) / df_exp['Men'][0]
s2 = ((df['Men'][1] - df_exp['Men'][1])**2) / df_exp['Men'][1]
s3 = ((df['Women'][0] - df_exp['Women'][0])**2) / df_exp['Women'][0]
s4 = ((df['Women'][1] - df_exp['Women'][1])**2) / df_exp['Women'][1]
Chi_sqr = round((s1+s2+s3+s4), 1) # rounding off till one decimal place
print('Computed value of Chi_Square is: ', Chi_sqr)

# Considering significace level ∝ = 5% = 0.05
# degree of freedom = (# of rows - 1) * (# of columns -1) = (2-1)*(2-1) =1

# prob = 1-0.05 = 0.95
p = 0.95
dof = 1

critical = chi2.ppf(p, dof)
print('\nCritical value of Chi-Square is: ', critical)

Observed table: 
            Men  Women
Voted      2792   3591
Not Voted  1486   2131 

Expected table: 
              Men   Women
Voted      2731.0  3652.0
Not Voted  1547.0  2070.0 

Computed value of Chi_Square is:  6.6

Critical value of Chi-Square is:  3.841458820694124


##### Conclusions
The criticial value of Chi-Square is: Χ^2 = 3.84 at 5% level of significance.
Our computed value of Chi-Square is: Χ^2 = 6.6

So, computed value falls in rejection region. Hence we reject the Null Hypothesis.
Hence gender and voting is not independent

### Problem Statement 19:  
 
A sample of 100 voters are asked which of four candidates they would vote for in an election. The number supporting each candidate is given below: 
 
 | Higgins| Reardon | White | Charlton|
| ---- | ---- | ---- | ---|
| 41 | 19| 24 | 16|

 
Do the data suggest that all candidates are equally popular? [Chi-Square = 14.96, with 3 df, 𝑝< 0.05]

In [21]:
from scipy.stats import chi2

# H0: All the candidates are equally popular.
# H1: The the candidates are not equally popular.

# In case of null hypothesis, the candidates will be equally popular if equal number of voters supports each candidate.
# frequency Expected = 100/4 = 25 for each candidate.

O = [41, 19, 24, 16] # Observed_frequency
E = 25 # Expected_frequency for each candidate

def chiSquare_Val(O, E):
    sum = 0
    for num in O:
        sum += (num-E)**2
    
    chi_val = sum / E
    return chi_val

X_sqr = chiSquare_Val(O, E)

print('Calculated Chi-Square value is: ', X_sqr)

# Given significace level ∝ = 5% = 0.05
# degree of freedom = 4-1 = 3

# prob = 1-0.05 = 0.95
p = 0.95
dof = 3

critical = chi2.ppf(p, dof)
print('Critical value of Chi-Square is: ', critical)

Calculated Chi-Square value is:  14.96
Critical value of Chi-Square is:  7.814727903251179


### Conclusions
    
The criticial value of Chi-Square is: Χ^2 = 7.81 at 5% level of significance. <br>
Our computed value of Chi-Square is: Χ^2 = 14.96 <br>

So, computed value falls in rejection region. Hence we reject the Null Hypothesis. <br>
Hence voters do not prefer the each candidate equally.

### Problem Statement 20: 
Children of three ages are asked to indicate their preference for three photographs of adults. Do the data suggest that there is a significant relationship between age and photograph preference? What is wrong with this study? [Chi-Square = 29.6, with 4 df: 𝑝  < 0.05].


| |  | | |Photograph|
| ---- | ---- | ---- | ---| ---|
|||A|B|C|
| Age of Child | 5-6 years| 18 | 22|20|
| | 7-8 years | 2 | 28|40|
||9-10 years|20|10|40|

In [22]:
import pandas as pd
import numpy as np
from scipy.stats import chi2

# creating view for Observed table

Photographs = {'A': [18, 2, 20], 'B': [22, 28, 10], 'C': [20, 40, 40]}
df_Obs = pd.DataFrame(Photographs, columns = ['A', 'B', 'C'], index = ['5-6 years', '7-8 years', '9-10 years'])
print('Observed table: ')
print(df_Obs , "\n")

# H0: No significant relationship between age of child and photograph preference.
# H1: There is significant relationship between age of child and photograph preference.

#Now we need to create expected table for the above null hypothesis. Formula -> E = ((row total * column total) / grand total)

Observed_lst = [18, 2, 20, 22, 28, 10, 20, 40, 40]

# Total for all combinations
grand_Total = sum(Observed_lst)

# Total 'A' / 'B' / 'C' for 5-6 years, 7-8 years and 9-10 years
T_col = df_Obs.sum(axis=0)

# Total 5-6 years / 7-8 years / 9-10 years
T_row = df_Obs.sum(axis=1)

# Now creating the expected list corresponding to the observed list

def exp_lst(T_col, T_row):
    Expected_lst = []
    for i in range(3):
        for j in range(3):
            res = (T_col[i] * T_row[j]) / grand_Total
            Expected_lst.append(res)
            
    return Expected_lst

Expected_lst = exp_lst(T_col, T_row)
print('The expected list is: ', Expected_lst)

# Now applying Chi-Square independence formula on Observed and Expected list

O_val = np.array(Observed_lst) # observed value
E_val = np.array(Expected_lst) # expected value

def chi_sqr(O_val, E_val):
    results = (O_val - E_val)**2 # (O-E)**2
    sum_sqr = 0
    for ele in range(len(results)):
        sum_sqr += (results[ele] / E_val[ele])
    return sum_sqr

X_sqr = chi_sqr(O_val, E_val)

print('Calculated Chi-Square value is: ', X_sqr)

# Given significace level ∝ = 5% = 0.05
# degree of freedom = (3-1) * (3-1) = 4

# prob = 1-0.05 = 0.95
p = 0.95
dof = 4

critical = chi2.ppf(p, dof)
print('Critical value of Chi-Square is: ', critical)
    

Observed table: 
             A   B   C
5-6 years   18  22  20
7-8 years    2  28  40
9-10 years  20  10  40 

The expected list is:  [12.0, 14.0, 14.0, 18.0, 21.0, 21.0, 30.0, 35.0, 35.0]
Calculated Chi-Square value is:  29.603174603174605
Critical value of Chi-Square is:  9.487729036781154


### Conclusions
    
The criticial value of Chi-Square is: Χ^2 = 9.488 at 5% level of significance. <br>
Our computed value of Chi-Square is: Χ^2 = 29.603 <br>

So, computed value falls in rejection region. Hence we reject the Null Hypothesis. <br>
Hence there is significant relationship between age of child and photograph preference.

### Problem Statement 21:  
 
A study of conformity using the Asch paradigm involved two conditions: one where one confederate supported the true judgement and another where no confederate gave the correct response.

 
| | Support | No Support | 
| ---- | ---- | ---- |
| Conform | 18| 40 |
| Not Conform| 32 | 10 |


Is there a significant difference between the "support" and "no support" conditions in the frequency with which individuals are likely to conform? [Chi-Square = 19.87, with 1 df: 𝑝  < 0.05]. 

In [23]:
import pandas as pd
import numpy as np
from scipy.stats import chi2

# creating view for Observed table

Asch_Para = {'Support': [18, 32], 'No Support': [40, 10]}
df_asch = pd.DataFrame(Asch_Para, columns = ['Support', 'No Support'], index = ['Conform', 'Not Conform'])
print('Observed table: ')
print(df_asch, "\n")

# H0: No significant difference between the 'support' and 'no support' conditions in the frequency.
# H1: There is significant difference between the 'support' and 'no support' conditions in the frequency.

#Now we need to create expected table for the above null hypothesis. Formula -> E = ((row total * column total) / grand total)

Observed_asch = [18, 32, 40, 10]

# Total for all combinations
grand_Total = sum(Observed_asch)

# Total 'support' / 'no support' for conform and not conform
T_columns = df_asch.sum(axis=0)

# Total conform and not conform
T_rows = df_asch.sum(axis=1)

# Now creating the expected list corresponding to the observed list

def exp_list(T_columns, T_rows):
    Expected_lst = []
    for i in range(2):
        for j in range(2):
            res = (T_columns[i] * T_rows[j]) / grand_Total
            Expected_lst.append(res)
            
    return Expected_lst

Expected_asch = exp_list(T_columns, T_rows)
print('The expected list is: ', Expected_asch)

# Now applying Chi-Square independence formula on Observed and Expected list

O_vals = np.array(Observed_asch) # observed value
E_vals = np.array(Expected_asch) # expected value

def chi_sqr(O_vals, E_vals):
    results = (O_vals - E_vals)**2 # (O-E)**2
    sum_sqr = 0
    for ele in range(len(results)):
        sum_sqr += (results[ele] / E_vals[ele])
    return sum_sqr

X_sqr = chi_sqr(O_vals, E_vals)

print('Calculated Chi-Square value is: ', X_sqr)

# Given significace level ∝ = 5% = 0.05
# degree of freedom = (2-1) * (2-1) = 1

# prob = 1-0.05 = 0.95
p = 0.95
dof = 1

critical = chi2.ppf(p, dof)
print('Critical value of Chi-Square is: ', critical)

Observed table: 
             Support  No Support
Conform           18          40
Not Conform       32          10 

The expected list is:  [29.0, 21.0, 29.0, 21.0]
Calculated Chi-Square value is:  19.868637110016422
Critical value of Chi-Square is:  3.841458820694124


### Conclusions
    
The criticial value of Chi-Square is: Χ^2 = 3.841 at 5% level of significance. <br>
Our computed value of Chi-Square is: Χ^2 = 19.87 <br>

So, computed value falls in rejection region. Hence we reject the Null Hypothesis. <br>
Hence there is significant difference between the 'support' and 'no support' conditions in the frequency.


### Problem Statement 22:  
 
We want to test whether short people differ with respect to their leadership qualities (Genghis Khan, Adolf Hitler and Napoleon were all stature-deprived, and how many midget MP's are there?) The following table shows the frequencies with which 43 short people and 52 tall people were categorized as "leaders", "followers" or as "unclassifiable". Is there a relationship between height and leadership qualities?  
[Chi-Square = 10.71, with 2 df: 𝑝  < 0.01]. 


| | Height |  |
| ---- | ---- | ---- |
|  | Short| Tall | 
| Leader | 12| 32 | 
| Follower| 22 | 14 |
|Unclassifiable|9|6|

In [24]:
import pandas as pd
import numpy as np
from scipy.stats import chi2

# creating view for Observed table

lead_tab = {'Short': [12, 22, 9], 'Tall': [32, 14, 6]}
df_frq = pd.DataFrame(lead_tab, columns = ['Short', 'Tall'], index = ['Leader', 'Follower', 'Unclassifiable'])
print('Observed table: ')
print(df_frq,"\n")

# H0: No relationship between height and leadership qualities.
# H1: There is relationship between height and leadership qualities.

#Now we need to create expected table for the above null hypothesis. Formula -> E = ((row total * column total) / grand total)

Obsv_lst = [12, 22, 9, 32, 14, 6]

# Total for all combinations
grand_Total = sum(Obsv_lst)

# Total 'short' / 'tall' column wise
To_cls = df_frq.sum(axis=0)

# Total leader / follower / unclassifiable row wise
To_rows = df_frq.sum(axis=1)

# Now creating the expected list corresponding to the observed list

def exp_list(To_cls, To_rows):
    Expected_lst = []
    for i in range(2):
        for j in range(3):
            res = (To_cls[i] * To_rows[j]) / grand_Total
            Expected_lst.append(res)
            
    return Expected_lst

Expects_lst = exp_list(To_cls, To_rows)
print('The expected list is: ', Expects_lst,"\n")

# Now applying Chi-Square independence formula on Observed and Expected list

Ob_vals = np.array(Obsv_lst) # observed value
Ex_vals = np.array(Expects_lst) # expected value

def chi_sqr(Ob_vals, Ex_vals):
    results = (Ob_vals - Ex_vals)**2 # (O-E)**2
    sum_sqr = 0
    for ele in range(len(results)):
        sum_sqr += (results[ele] / Ex_vals[ele])
    return sum_sqr

X_sqr = chi_sqr(Ob_vals, Ex_vals)

print('Calculated Chi-Square value is: ', X_sqr)

# Given significace level ∝ = 1% = 0.01
# degree of freedom = (2-1) * (3-1) = 2

# prob = 1-0.01 = 0.99
p = 0.99
dof = 2

critical = chi2.ppf(p, dof)
print('Critical value of Chi-Square is: ', critical)

Observed table: 
                Short  Tall
Leader             12    32
Follower           22    14
Unclassifiable      9     6 

The expected list is:  [19.91578947368421, 16.294736842105262, 6.7894736842105265, 24.08421052631579, 19.705263157894738, 8.210526315789474] 

Calculated Chi-Square value is:  10.712198008709638
Critical value of Chi-Square is:  9.21034037197618


### Conclusions
    
The criticial value of Chi-Square is: Χ^2 = 9.21 at 1% level of significance. <br>
Our computed value of Chi-Square is: Χ^2 = 10.71 <br>

So, computed value falls in rejection region. Hence we reject the Null Hypothesis. <br>
Hence there is relationship between height and leadership qualitie

#### Problem Statement 23:  
 
Each respondent in the Current Population Survey of March 1993 was classified as employed, unemployed, or outside the labor force. The results for men in California age 35 to 44 can be cross-tabulated by marital status, as follows: 

| | Married | Widow, divorced or separated | Never Married|
| ---- | ---- | ---- | ---|
| Employed | 679| 103 | 114|
| Unemployed| 63 | 10 | 20|
|Not in labor force|42|18|25|

Men of different marital status seem to have different distributions of labor force status. Or is this just chance variation? (you may assume the table results from a simple random sample.) 

In [25]:
import pandas as pd
import numpy as np
from scipy.stats import chi2

# creating view for Observed table

population = {'Married': [679, 63, 42], 'Widow, divorced or separated': [103, 10, 18], 'Never Married' : [114,20,25]}
df_surv = pd.DataFrame(population, columns = ['Married', 'Widow, divorced or separated', 'Never Married'], index = ['Employed', 'Unemployed', 'Not in labor force'])
print('Observed table: ')
print(df_surv,"\n")

# H0: No relationship between men of different marital status and different distribution of labor force status.
# H1: There is relationship between men of different marital status and different distribution of labor force status.

#Now we need to create expected table for the above null hypothesis. Formula -> E = ((row total * column total) / grand total)

Obsv = [679, 63, 42, 103, 10, 18, 114, 20, 25]

# Total for all combinations
grand_Total = sum(Obsv)

# Total 'married' / 'widow ...' / 'never married' column wise
sum_cols = df_surv.sum(axis=0)

# Total employed / unemployed / not in labor force row wise
sum_rows = df_surv.sum(axis=1)

# Now creating the expected list corresponding to the observed list

def exp_lsts(sum_cols, sum_rows):
    Expected_lst = []
    for i in range(3):
        for j in range(3):
            res = (sum_cols[i] * sum_rows[j]) / grand_Total
            Expected_lst.append(res)
            
    return Expected_lst

expected_res = exp_lsts(sum_cols, sum_rows)
print('The expected list is: ', expected_res, "\n")

# Now applying Chi-Square independence formula on Observed and Expected list

Obs_vals = np.array(Obsv) # observed value
Exp_vals = np.array(expected_res) # expected value

def chi_sqr(Obs_vals, Exp_vals):
    results = (Obs_vals - Exp_vals)**2 # (O-E)**2
    sum_sqr = 0
    for ele in range(len(results)):
        sum_sqr += (results[ele] / Exp_vals[ele])
    return sum_sqr

X_sqre = chi_sqr(Obs_vals, Exp_vals)

print('Calculated Chi-Square value is: ', X_sqre)

# Given significace level ∝ = 5% = 0.05
# degree of freedom = (3-1) * (3-1) = 4

# prob = 1-0.05 = 0.95
p = 0.95
dof = 4

critical = chi2.ppf(p, dof)
print('Critical value of Chi-Square is: ', critical)

Observed table: 
                    Married  Widow, divorced or separated  Never Married
Employed                679                           103            114
Unemployed               63                            10             20
Not in labor force       42                            18             25 

The expected list is:  [654.0633147113595, 67.88826815642459, 62.04841713221602, 109.28864059590316, 11.343575418994414, 10.36778398510242, 132.64804469273744, 13.768156424581006, 12.583798882681565] 

Calculated Chi-Square value is:  31.61310319407798
Critical value of Chi-Square is:  9.487729036781154


#### Conclusions
The criticial value of Chi-Square is: Χ^2 = 9.49 at 5% level of significance.
Our computed value of Chi-Square is: Χ^2 = 31.61

So, computed value falls in rejection region. Hence we reject the Null Hypothesis.
Hence there is relationship between men of different marital status and different distribution of labor force status